Mathlib/Geometry/Manifold/LocalDiffeomorph.lean

```diff
@@ -0,0 +1,246 @@
+/-- A partial diffeomorphism on `s` is a function `f : M → N` such that `f` restricts to a
+diffeomorphism `s → t` between open subsets of `M` and `N`, respectively.
+This is an auxiliary definition and should not be used outside of this file. -/
+structure PartialDiffeomorph extends PartialEquiv M N where
+  open_source : IsOpen source
+  open_target : IsOpen target
+  contMDiffOn_toFun : ContMDiffOn I J n toFun source
+  contMDiffOn_invFun : ContMDiffOn J I n invFun target
+
+/-- Coercion of a `PartialDiffeomorph` to function.
+Note that a `PartialDiffeomorph` is not `FunLike` (like `PartialHomeomorph`),
+as `toFun` doesn't determine `invFun` outside of `target`. -/
+instance : CoeFun (PartialDiffeomorph I J M N n) fun _ => M → N :=
+  ⟨fun Φ => Φ.toFun⟩
+
+-- Add the very basic API we need.
+namespace PartialDiffeomorph
+variable (Φ : PartialDiffeomorph I J M N n) (hn : 1 ≤ n)
+
+/-- A partial diffeomorphism is also a local homeomorphism. -/
+def toPartialHomeomorph : PartialHomeomorph M N :=
+  {
+    toPartialEquiv := Φ.toPartialEquiv
+    open_source := Φ.open_source
+    open_target := Φ.open_target
+    continuousOn_toFun := Φ.contMDiffOn_toFun.continuousOn
+    continuousOn_invFun := Φ.contMDiffOn_invFun.continuousOn
+  }
+
+/-- The inverse of a local diffeomorphism. -/
+protected def symm : PartialDiffeomorph J I N M n :=
+  {
+    toPartialEquiv := Φ.toPartialEquiv.symm
+    open_source := Φ.open_target
+    open_target := Φ.open_source
+    contMDiffOn_toFun := Φ.contMDiffOn_invFun
+    contMDiffOn_invFun := Φ.contMDiffOn_toFun
+  }
+
+protected theorem contMDiffOn : ContMDiffOn I J n Φ Φ.source :=
+  Φ.contMDiffOn_toFun
+
+protected theorem contMDiffOn_symm : ContMDiffOn J I n Φ.symm Φ.target :=
+  Φ.contMDiffOn_invFun
+
+protected theorem mdifferentiableOn : MDifferentiableOn I J Φ Φ.source :=
+  (Φ.contMDiffOn).mdifferentiableOn hn
+
+protected theorem mdifferentiableOn_symm : MDifferentiableOn J I Φ.symm Φ.target :=
+  (Φ.symm).mdifferentiableOn hn
+
+protected theorem mdifferentiableAt {x : M} (hx : x ∈ Φ.source) : MDifferentiableAt I J Φ x :=
+  (Φ.mdifferentiableOn hn x hx).mdifferentiableAt (Φ.open_source.mem_nhds hx)
+
+protected theorem mdifferentiableAt_symm {x : M} (hx : x ∈ Φ.source) :
+    MDifferentiableAt J I Φ.symm (Φ x) :=
+  (Φ.symm).mdifferentiableAt hn (Φ.map_source hx)
+
+/- We could add lots of additional API (following `Diffeomorph` and `PartialHomeomorph`), such as
+* further continuity and differentiability lemmas
+* refl and trans instances; lemmas between them.
+As this declaration is meant for internal use only, we keep it simple. -/
+end PartialDiffeomorph
+end PartialDiffeomorph
+
+variable {M N}
+
+/-- `f : M → N` is called a **`C^n` local diffeomorphism at *x*** iff there exist
+  open sets `U ∋ x` and `V ∋ f x` and a diffeomorphism `Φ : U → V` such that `f = Φ` on `U`. -/
+def IsLocalDiffeomorphAt (f : M → N) (x : M) : Prop :=
+  ∃ Φ : PartialDiffeomorph I J M N n, x ∈ Φ.source ∧ EqOn f Φ Φ.source
+
```

Comment:
You're missing `IsLocalDiffeomorphOn`. In all the examples I can think of, when we didn't define the `On` version to start with, we have regretted it later on and we had to refactor things. So, would you agree to add it from the start? (I know it's more work, of course...)

Reply:
I was wondering about this :-) I'll trust your collective experience, I'll add it.